os/drivers/virtkey.c

```diff
@@ -0,0 +1,320 @@
+			priv->events[priv->tail] = buffer[i];
+			priv->tail = (priv->tail + 1) % MAX_LEN;
+			priv->size++;
+			ret++;
+		}
+	}
+
+	virtkey_givesem(&priv->datasem);
+
+	if (ret > 0) {
+		/* notify only if we got some data */
+		virtkey_pollnotify(priv, POLLIN);
+	}
+
+	return ret;
+}
+
+#ifndef CONFIG_DISABLE_POLL
+static void virtkey_pollnotify(FAR struct virtkey_dev_s *dev, pollevent_t eventset)
+{
+	int i;
+	for (i = 0; i < CONFIG_SERIAL_NPOLLWAITERS; i++) {
+		struct pollfd *fds = dev->fds[i];
+		if (fds) {
+			fds->revents |= (fds->events & eventset);
+			if (fds->revents != 0) {
+				sem_post(fds->sem);
+			}
+		}
+	}
+}
+#else
+#define virtkey_pollnotify(dev, event)
+#endif
+
+#ifndef CONFIG_DISABLE_POLL
+int virtkey_poll(FAR struct file *filep, FAR struct pollfd *fds, bool setup)
+{
+	FAR struct inode *inode = filep->f_inode;
+	FAR struct virtkey_dev_s *dev = inode->i_private;
+	pollevent_t eventset;
+	int ret;
+	int i;
+
+	/* Some sanity checking */
+
+#ifdef CONFIG_DEBUG
+	if (!dev || !fds) {
+		return -ENODEV;
+	}
+#endif
+
+	/* Are we setting up the poll?  Or tearing it down? */
+
+	ret = virtkey_takesem(&dev->pollsem, true);
+	if (ret < 0) {
+		/* A signal received while waiting for access to the poll data
+		 * will abort the operation.
+		 */
+
+		return ret;
+	}
+
+	if (setup) {
+		/* This is a request to set up the poll.  Find an available
+		 * slot for the poll structure reference
+		 */
+
+		for (i = 0; i < CONFIG_SERIAL_NPOLLWAITERS; i++) {
+			/* Find an available slot */
+
+			if (!dev->fds[i]) {
+				/* Bind the poll structure and this slot */
+
+				dev->fds[i] = fds;
+				fds->priv = &dev->fds[i];
+				fds->filep = (void *)filep;
+				break;
+			}
+		}
+
+		if (i >= CONFIG_SERIAL_NPOLLWAITERS) {
+			fds->priv = NULL;
+			fds->filep = NULL;
+			ret = -EBUSY;
+			goto errout;
+		}
+
+		eventset = 0;
+		(void)virtkey_takesem(&dev->datasem, false);
+
+		/* use poll only for reading events */
+		if (dev->tail != dev->head) {
+			eventset |= (fds->events & POLLIN);
+		}
+
+		virtkey_givesem(&dev->datasem);
+
+		if (eventset) {
+			virtkey_pollnotify(dev, eventset);
+		}
+
+	} else if (fds->priv) {
+		/* This is a request to tear down the poll. */
+
+		struct pollfd **slot = (struct pollfd **)fds->priv;
+
+		/* Remove all memory of the poll setup */
+
+		*slot = NULL;
+		fds->priv = NULL;
+		fds->filep = NULL;
+	}
+
+errout:
+	virtkey_givesem(&dev->pollsem);
+	return ret;
+}
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: virtkey_register
+ *
+ * Description:
+ *   Register /dev/virktey
+ *
+ ****************************************************************************/
+
+void virtkey_register(void)
+{
+	struct virtkey_dev_s *priv = &g_virtkey_priv;
+	
+	priv->events = (char *) malloc(MAX_LEN);
```

Comment:
1. Let's remove a space after type casting.
2. Is it `malloc`? not `kmm_malloc`?